README.md
# Karl

Plays Kaput (Farkle with full straight as 2000).


changelog.md
# Changelog & Scores

## VERSION 0.0.1

**Won: 0**
**Lost: 1**

Beat Evan!

Optimal play:
```
player.rerollThreshold = 5
player.stingyFives = True
player.rollAfterFull = True
player.rerollThresholdAfterFull = 5
```

## Version 0.2.0

* Added option to ignore triple 2's
* Added option to abandon 5 collection at a certain threshold
* Added endgame mode management
* Process 1's and 5's separately

**Won: 0**
**Lost: 1**


Optimal play:
```
player.stingyFives = True
player.rollAfterFull = True
player.ignoreTriple2 = True
player.rerollThreshold = 6
player.rerollThresholdAfterFull = 6
player.abandon5Threshold = 450
```


testing/testing.py
import hand
import numpy as np
import matplotlib.pyplot as plt
import pprint
import copy

# handCounts = [10, 100, 1000, 10000, 100000]
handCounts = [100, 1000, 10000]
debugOutput = False
lastPercentage = -1

for handCount in handCounts:
    bestMHTW = 99999.99
    bestMHTWObj = hand.kaputHand()

    print("Simulating " + str(handCount) + " hands per tuning option; sit tight")
    print("==============================")

    testHands = []
    gameModes = 0
    gameModesPlayed = 0

    # mode assembly
    for rollAfterFullIter in range(2):
        options = {}
        if rollAfterFullIter == 0:
            options['rollAfterFull'] = False
        else:
            options['rollAfterFull'] = True

        for stingyFivesIter in range(2):
            if stingyFivesIter == 0:
                options['stingyFives'] = False
            else:
                options['stingyFives'] = True

            for ignoreTriple2Iter in range(2):
                if ignoreTriple2Iter == 0:
                    options['ignoreTriple2'] = False
                else:
                    options['ignoreTriple2'] = True

                for rerollThresholdAfterFullSet in range(6):
                    options['rerollThresholdAfterFull'] = rerollThresholdAfterFullSet + 1

                    for rerollThresholdSet in range(6):
                        options['rerollThreshold'] = rerollThresholdSet + 1

                        for abandon5ThresholdSet in range(50, 500, 50):
                            options['abandon5Threshold'] = abandon5ThresholdSet

                            testHands.append(options.copy())
                            gameModes += 1

    # mode testing
    for testHand in testHands:
        testHandObj = hand.kaputHand()
        testHandObj.silent = True

        testHandObj.stingyFives = testHand['stingyFives']
        testHandObj.rollAfterFull = testHand['rollAfterFull']
        testHandObj.rerollThreshold = testHand['rerollThreshold']
        testHandObj.rerollThresholdAfterFull = testHand['rerollThresholdAfterFull']
        testHandObj.ignoreTriple2 = testHand['ignoreTriple2']
        testHandObj.abandon5Threshold = testHand['abandon5Threshold']

        if debugOutput:
            print("stingyFives: " + str(testHandObj.stingyFives))
            print("rollAfterFull: " + str(testHandObj.rollAfterFull))
            print("ignoreTriple2: " + str(testHandObj.ignoreTriple2))
            print("rerollThreshold: " + str(testHandObj.rerollThreshold))
            print("rerollThresholdAfterFull: " + str(testHandObj.rerollThresholdAfterFull))
            print("abandon5Threshold: " + str(testHandObj.abandon5Threshold))

        turnsTo5K = []

        for _ in range(handCount):
            testHandObj.refresh()
            score = 0
            turnCount = 0

            while score < 5000:
                turnCount += 1
                testHandObj.refresh()
                while testHandObj.handInPlay:
                    testHandObj.roll()
                score += testHandObj.score

            turnsTo5K.append(turnCount)

        results = np.array(turnsTo5K)
        mhtw = np.mean(results)
        gameModesPlayed += 1
        # print("Mean hands to win: " +
        #       str(mhtw))
        if round((gameModesPlayed / gameModes) * 100) % 5 == 0 and round((gameModesPlayed / gameModes) * 100) > lastPercentage:
            print("Played " + str(gameModesPlayed) + " modes out of " + str(gameModes) + "(" + str(round((gameModesPlayed / gameModes) * 100)) + "% done)")
            lastPercentage = round((gameModesPlayed / gameModes) * 100)

        # handle best scoring
        if mhtw.item() < bestMHTW:
            bestMHTW = mhtw
            bestMHTWObj = copy.deepcopy(testHandObj)

    print("====== OPTIMUM TUNING: " + str(bestMHTW) + " MHTW ======")
    print("stingyFives: " + str(testHandObj.stingyFives))
    print("rollAfterFull: " + str(testHandObj.rollAfterFull))
    print("ignoreTriple2: " + str(testHandObj.ignoreTriple2))
    print("rerollThreshold: " + str(testHandObj.rerollThreshold))
    print("rerollThresholdAfterFull: " + str(testHandObj.rerollThresholdAfterFull))
    print("abandon5Threshold: " + str(testHandObj.abandon5Threshold))
    print("Played " + str(handCount * gameModes) + " games in " + str(gameModes) + " modes")

    timeTo5K = []
    for _ in range(handCount):
        bestMHTWObj.refresh()
        score = 0
        turnCount = 0

        while score < 5000:
            turnCount += 1
            bestMHTWObj.refresh()
            while bestMHTWObj.handInPlay:
                bestMHTWObj.roll()
            score += bestMHTWObj.score

        timeTo5K.append(turnCount)

    timeTo5KNP = np.array(timeTo5K)

    print("Percentiles:")
    print("\t10%: " + str(np.percentile(timeTo5KNP, 10)))
    print("\t50%: " + str(np.percentile(timeTo5KNP, 50)))
    print("\t90%: " + str(np.percentile(timeTo5KNP, 90)))
    print("\t95%: " + str(np.percentile(timeTo5KNP, 95)))
    print("\t99%: " + str(np.percentile(timeTo5KNP, 99)))

    plt.hist(timeTo5KNP, bins=range(1, 20))
    plt.title("MHTW: " + str(bestMHTW) + " -- " + str(handCount) + " iterations")
    plt.xlabel("Mean Hands to Win")
    plt.ylabel("Occurrences")
    plt.savefig("optimum-test-" + str(handCount) + ".png")
    print("Histogram saved!")


testing/testing_singlemode.py
import hand
import numpy as np
import matplotlib.pyplot as plt

handCount = 1100
turnsTo5K = []

player = hand.kaputHand()
player.silent = True

player.stingyFives = True
player.rollAfterFull = True
player.ignoreTriple2 = False
player.rerollThreshold = 4
player.rerollThresholdAfterFull = 4
player.abandon5Threshold = 50

for _ in range(handCount):
    player.refresh()
    score = 0
    turnCount = 0

    while score < 5000:
        turnCount += 1
        player.refresh()
        while player.handInPlay:
            player.roll()
        score += player.score

    turnsTo5K.append(turnCount)

results = np.array(turnsTo5K)
mean = np.mean(results)

print("====== SINGLE STRATEGY DRILLDOWN: " + str(mean) + " MHTW ======")
print("stingyFives: " + str(player.stingyFives))
print("rollAfterFull: " + str(player.rollAfterFull))
print("ignoreTriple2: " + str(player.ignoreTriple2))
print("rerollThreshold: " + str(player.rerollThreshold))
print("rerollThresholdAfterFull: " + str(player.rerollThresholdAfterFull))
print("abandon5Threshold: " + str(player.abandon5Threshold))
print("")

print("Percentiles & Mean:")
print("\tMean: " + str(mean))
print("\t10%: " + str(np.percentile(results, 10)))
print("\t50%: " + str(np.percentile(results, 50)))
print("\t90%: " + str(np.percentile(results, 90)))
print("\t95%: " + str(np.percentile(results, 95)))
print("\t99%: " + str(np.percentile(results, 99)))

plt.hist(results, bins=range(1, 20))
plt.title("MHTW " + str(mean) + " -- " + str(handCount) + " iterations")
plt.xlabel("Mean Hands to Win")
plt.ylabel("Occurrences")
# plt.show()


hand.py
import random
import numpy

class kaputHand:
    TOTALDIE = 6
    STRAIGHTSCORE = 2000

    # DATA STORAGE
    rolled = []  # the hand that was rolled
    kept = []  # the currently retained hand
    rolls = 0  # number of rolls
    fulls = 0  # the total number of fulls
    score = 0  # total score
    kaput = False  # if the player has gone kaput
    handInPlay = True  # if the player is done
    scoreToBeat = 0  # endgame mode score to beat

    # OUTPUT
    printDebugOutput = True  # print debug output each roll
    silent = False  # print no output at all

    # GAME MODES
    endGameMode = False
    interactivePlay = False

    # TUNING
    stingyFives = True  # if there are multiple fives and no ones in a vanilla roll (no straight), take one and reroll
    rollAfterFull = True  # should the bot roll after recieving a flush
    ignoreTriple2 = False  # ignore 222 as 200
    rerollThreshold = 4  # number of dice left (or more) to reroll once 300 is reached
    rerollThresholdAfterFull = 5  # after a full, what should the rerollThreshold be
    abandon5Threshold = 300  # the score at which, assuming 1 die has been retained, Karl should stop taking fives

    def output(self, string):
        if not self.silent:
            print(string)

    def refresh(self):
        self.rolled = []
        self.kept = []
        self.rolls = 0
        self.fulls = 0
        self.score = 0
        self.kaput = False
        self.handInPlay = True
        self.endGameMode = False
        self.scoreToBeat = 0

    def diceToRoll(self):
        return self.TOTALDIE - len(self.kept)

    def shouldRoll(self):
        if self.endGameMode:
            if self.score <= self.scoreToBeat:
                return True
            else:
                return False

        if self.fulls > 0:
            if not self.rollAfterFull:
                return False
            if self.diceToRoll() >= self.rerollThresholdAfterFull:
                return True

        if self.score < 300:
            return True

        if self.diceToRoll() >= self.rerollThreshold:
            return True

        return False

    def getRoll(self):
        if self.endGameMode and self.scoreToBeat == 0:
            self.scoreToBeat = int(input("Score to beat: "))

        if self.interactivePlay:
            prompt = "Enter roll (" + str(self.diceToRoll()) + " dice, or 'e' to enter endgame mode): "
            if self.endGameMode:
                prompt = "Enter roll (" + str(self.diceToRoll()) + " dice, in endgame mode -- " + str(self.scoreToBeat) + " to beat): "

            inputRoll = input(prompt)

            # convert to endgame mode
            if "e" in inputRoll:
                self.endGameMode = True
                return self.getRoll()

            return sorted([int(i) for i in list(inputRoll)])
        else:
            return sorted([numpy.random.randint(1, 7) for
                          i in range(self.diceToRoll())])

    def roll(self):
        if self.shouldRoll():
            self.rolled = self.getRoll()
            self.output("ROLLED: " + str(self.rolled))
            self.rolls += 1
            self.checkRoll()
        else:
            self.output("Ending turn.")
            self.handInPlay = False

        self.printStatus()

    def checkRoll(self):
        keptThisHand = 0

        # check straight
        if self.rolled == [1, 2, 3, 4, 5, 6]:
            self.score += self.STRAIGHTSCORE
            self.fulls += 1
            self.rolls = 0
            return

        # find triples
        triples = set([x for x in self.rolled if self.rolled.count(x) >= 3])
        for value in triples:
            # 2s handling
            if value == 2:
                if self.ignoreTriple2:
                    # only ignore 2's if we can
                    if self.rolled.count(1) > 0 or self.rolled.count(5) > 0 or len(triples) > 1:
                        continue

            if value == 1:
                self.score += 1000
            else:
                self.score += value * 100  # take the score

            for _ in range(3):
                self.rolled.remove(value)  # pop the die
                self.kept.append(value)

            keptThisHand += 3

        # handle multiple 50s where fifties don't make the whole roll
        # and there are no 1s
        noOnes = self.rolled.count(5) > 1 and self.rolled.count(1) == 0
        onlyFives = (len(self.rolled) - self.rolled.count(5)) == 0
        if noOnes and not onlyFives and self.stingyFives:
            self.score += 50
            self.kept.append(5)
            self.rolled.remove(5)
            return

        iterableRolled = self.rolled[:]
        # do basic score count
        for value in iterableRolled:
            if value == 1:
                self.score += 100
                self.kept.append(value)
                self.rolled.remove(value)
                keptThisHand += 1

        for value in iterableRolled:
            if value == 5:
                if keptThisHand >= 1 and self.score > self.abandon5Threshold:
                    return
                else:
                    self.score += 50
                    self.kept.append(value)
                    self.rolled.remove(value)
                    keptThisHand += 1

        if keptThisHand == 0:
            self.handInPlay = False
            self.score = 0
            self.kaput = True

        # handle full
        if len(self.kept) == 6:
            self.fulls += 1
            self.kept = []
        else:
            if self.rolls == 3 and not self.endGameMode:
                self.handInPlay = False
                if self.score < 300:
                    self.score = 0
                    self.kaput = True

    def debugStatus(self):
        print("Roll Debug")
        print("\tRoll count: " + str(self.rolls))
        print("\tFulls: " + str(self.fulls))
        print("\tCurrent kept: " + str(self.kept))
        print("\tCurrent score: " + str(self.score))
        print("\tHand in play? " + str(self.handInPlay))
        print("\tKaput? " + str(self.kaput))
        print("")

    def endTurnStatus(self):
        print("Turn status:")
        print("\tKept: " + str(self.kept))
        print("\tScore: " + str(self.score))
        if self.fulls > 0:
            print("\tFulls: " + str(self.fulls))
        print("\tKaput? " + str(self.kaput))
        print("")

    def printStatus(self):
        if not self.silent:
            if self.handInPlay and self.printDebugOutput:
                self.debugStatus()
            else:
                self.endTurnStatus()


testing_singlemode.py
import hand
import numpy as np
import matplotlib.pyplot as plt

handCount = 1100
turnsTo5K = []

player = hand.kaputHand()
player.silent = True

player.stingyFives = True
player.rollAfterFull = True
player.ignoreTriple2 = False
player.rerollThreshold = 4
player.rerollThresholdAfterFull = 4
player.abandon5Threshold = 450

for _ in range(handCount):
    player.refresh()
    score = 0
    turnCount = 0

    while score < 5000:
        turnCount += 1
        player.refresh()
        while player.handInPlay:
            player.roll()
        score += player.score

    turnsTo5K.append(turnCount)

results = np.array(turnsTo5K)
mean = np.mean(results)

print("====== SINGLE STRATEGY DRILLDOWN: " + str(mean) + " MHTW ======")
print("stingyFives: " + str(player.stingyFives))
print("rollAfterFull: " + str(player.rollAfterFull))
print("ignoreTriple2: " + str(player.ignoreTriple2))
print("rerollThreshold: " + str(player.rerollThreshold))
print("rerollThresholdAfterFull: " + str(player.rerollThresholdAfterFull))
print("abandon5Threshold: " + str(player.abandon5Threshold))
print("")

print("Percentiles & Mean:")
print("\tMean: " + str(mean))
print("\tp1: " + str(np.percentile(results, 1)))
print("\tp25: " + str(np.percentile(results, 10)))
print("\tp50: " + str(np.percentile(results, 50)))
print("\tp75: " + str(np.percentile(results, 50)))
print("\tp90: " + str(np.percentile(results, 90)))

plt.hist(results, bins=range(1, 20))
plt.title("MHTW " + str(mean) + " -- " + str(handCount) + " iterations")
plt.xlabel("Mean Hands to Win")
plt.ylabel("Occurrences")
# plt.show()


testing.py
import hand
import numpy as np
import matplotlib.pyplot as plt
import pprint
import copy

# handCounts = [10, 100, 1000, 10000, 100000]
handCounts = [100, 1000, 10000]
debugOutput = False
lastPercentage = -1

for handCount in handCounts:
    bestMHTW = 99999.99
    bestMHTWObj = hand.kaputHand()

    print("Simulating " + str(handCount) + " hands per tuning option; sit tight")
    print("==============================")

    testHands = []
    gameModes = 0
    gameModesPlayed = 0

    # mode assembly
    for rollAfterFullIter in range(2):
        options = {}
        if rollAfterFullIter == 0:
            options['rollAfterFull'] = False
        else:
            options['rollAfterFull'] = True

        for stingyFivesIter in range(2):
            if stingyFivesIter == 0:
                options['stingyFives'] = False
            else:
                options['stingyFives'] = True

            for ignoreTriple2Iter in range(2):
                if ignoreTriple2Iter == 0:
                    options['ignoreTriple2'] = False
                else:
                    options['ignoreTriple2'] = True

                for rerollThresholdAfterFullSet in range(6):
                    options['rerollThresholdAfterFull'] = rerollThresholdAfterFullSet + 1

                    for rerollThresholdSet in range(6):
                        options['rerollThreshold'] = rerollThresholdSet + 1

                        for abandon5ThresholdSet in range(50, 500, 50):
                            options['abandon5Threshold'] = abandon5ThresholdSet

                            testHands.append(options.copy())
                            gameModes += 1

    # mode testing
    for testHand in testHands:
        testHandObj = hand.kaputHand()
        testHandObj.silent = True

        testHandObj.stingyFives = testHand['stingyFives']
        testHandObj.rollAfterFull = testHand['rollAfterFull']
        testHandObj.rerollThreshold = testHand['rerollThreshold']
        testHandObj.rerollThresholdAfterFull = testHand['rerollThresholdAfterFull']
        testHandObj.ignoreTriple2 = testHand['ignoreTriple2']
        testHandObj.abandon5Threshold = testHand['abandon5Threshold']

        if debugOutput:
            print("stingyFives: " + str(testHandObj.stingyFives))
            print("rollAfterFull: " + str(testHandObj.rollAfterFull))
            print("ignoreTriple2: " + str(testHandObj.ignoreTriple2))
            print("rerollThreshold: " + str(testHandObj.rerollThreshold))
            print("rerollThresholdAfterFull: " + str(testHandObj.rerollThresholdAfterFull))
            print("abandon5Threshold: " + str(testHandObj.abandon5Threshold))

        turnsTo5K = []

        for _ in range(handCount):
            testHandObj.refresh()
            score = 0
            turnCount = 0

            while score < 5000:
                turnCount += 1
                testHandObj.refresh()
                while testHandObj.handInPlay:
                    testHandObj.roll()
                score += testHandObj.score

            turnsTo5K.append(turnCount)

        results = np.array(turnsTo5K)
        mhtw = np.mean(results)
        gameModesPlayed += 1
        # print("Mean hands to win: " +
        #       str(mhtw))
        if round((gameModesPlayed / gameModes) * 100) % 5 == 0 and round((gameModesPlayed / gameModes) * 100) > lastPercentage:
            print("Played " + str(gameModesPlayed) + " modes out of " + str(gameModes) + "(" + str(round((gameModesPlayed / gameModes) * 100)) + "% done)")
            lastPercentage = round((gameModesPlayed / gameModes) * 100)

        # handle best scoring
        if mhtw.item() < bestMHTW:
            bestMHTW = mhtw
            bestMHTWObj = copy.deepcopy(testHandObj)

    print("====== OPTIMUM TUNING: " + str(bestMHTW) + " MHTW ======")
    print("stingyFives: " + str(testHandObj.stingyFives))
    print("rollAfterFull: " + str(testHandObj.rollAfterFull))
    print("ignoreTriple2: " + str(testHandObj.ignoreTriple2))
    print("rerollThreshold: " + str(testHandObj.rerollThreshold))
    print("rerollThresholdAfterFull: " + str(testHandObj.rerollThresholdAfterFull))
    print("abandon5Threshold: " + str(testHandObj.abandon5Threshold))
    print("Played " + str(handCount * gameModes) + " games in " + str(gameModes) + " modes")

    timeTo5K = []
    for _ in range(handCount):
        bestMHTWObj.refresh()
        score = 0
        turnCount = 0

        while score < 5000:
            turnCount += 1
            bestMHTWObj.refresh()
            while bestMHTWObj.handInPlay:
                bestMHTWObj.roll()
            score += bestMHTWObj.score

        timeTo5K.append(turnCount)

    timeTo5KNP = np.array(timeTo5K)

    print("Percentiles & Mean:")
    print("\tMean: " + str(bestMHTW))
    print("\tp1: " + str(np.percentile(timeTo5KNP, 1)))
    print("\tp25: " + str(np.percentile(timeTo5KNP, 10)))
    print("\tp50: " + str(np.percentile(timeTo5KNP, 50)))
    print("\tp75: " + str(np.percentile(timeTo5KNP, 50)))
    print("\tp90: " + str(np.percentile(timeTo5KNP, 90)))

    plt.hist(timeTo5KNP, bins=range(1, 20))
    plt.title("MHTW: " + str(bestMHTW) + " -- " + str(handCount) + " iterations")
    plt.xlabel("Mean Hands to Win")
    plt.ylabel("Occurrences")
    plt.savefig("optimum-test-" + str(handCount) + ".png")
    print("Histogram saved!")
